File: jambeth/jambeth-cache/src/main/java/de/osthus/ambeth/util/AlreadyHandledSet.java
package de.osthus.ambeth.util;

import de.osthus.ambeth.collections.Tuple2KeyEntry;
import de.osthus.ambeth.collections.Tuple2KeyHashMap;

public class AlreadyHandledSet extends Tuple2KeyHashMap<Object, PrefetchPath[], Boolean>
{
	@Override
	protected boolean equalKeys(Object obj, PrefetchPath[] prefetchPaths, Tuple2KeyEntry<Object, PrefetchPath[], Boolean> entry)
	{
		return obj == entry.getKey1() && prefetchPaths == entry.getKey2();
	}

	@Override
	protected int extractHash(Object obj, PrefetchPath[] prefetchPaths)
	{
		if (prefetchPaths == null)
		{
			return System.identityHashCode(obj);
		}
		return System.identityHashCode(obj) ^ prefetchPaths.hashCode();
	}
}

File: ambeth/Ambeth.Cache/ambeth/util/AlreadyHandledSet.cs
using De.Osthus.Ambeth.Collections;
using System;
using System.Runtime.CompilerServices;

namespace De.Osthus.Ambeth.Util
{
	public class AlreadyHandledSet : Tuple2KeyHashMap<Object, PrefetchPath[], bool?>
	{
		protected override bool EqualKeys(Object obj, PrefetchPath[] prefetchPaths, Tuple2KeyEntry<Object, PrefetchPath[], bool?> entry)
		{
			return obj == entry.GetKey1() && prefetchPaths == entry.GetKey2();
		}

		protected override int ExtractHash(Object obj, PrefetchPath[] prefetchPaths)
		{
			if (prefetchPaths == null)
			{
				return RuntimeHelpers.GetHashCode(obj);
			}
			return RuntimeHelpers.GetHashCode(obj) ^ prefetchPaths.GetHashCode();
		}
	}
}